tests/network-tests/src/nicaea/tests/proposals/manageLeaderRole.ts

```diff
@@ -0,0 +1,213 @@
+import { initConfig } from '../../utils/config';
+import { Keyring, WsProvider } from '@polkadot/api';
+import BN from 'bn.js';
+import { setTestTimeout } from '../../utils/setTestTimeout';
+import tap from 'tap';
+import { registerJoystreamTypes } from '@nicaea/types';
+import { closeApi } from '../impl/closeApi';
+import { ApiWrapper, WorkingGroups } from '../../utils/apiWrapper';
+import {
+  createWorkingGroupLeaderOpening,
+  voteForProposal,
+  beginWorkingGroupLeaderApplicationReview,
+  fillLeaderOpeningProposal,
+  terminateLeaderRoleProposal,
+  setLeaderRewardProposal,
+  decreaseLeaderStakeProposal,
+  slashLeaderProposal,
+} from './impl/proposalsModule';
+import {
+  applyForOpening,
+  expectLeadOpeningAdded,
+  expectLeaderSet,
+  expectBeganApplicationReview,
+  expectLeaderRoleTerminated,
+  expectLeaderRewardAmountUpdated,
+  expectLeaderStakeDecreased,
+  expectLeaderSlashed,
+} from '../workingGroup/impl/workingGroupModule';
+
+tap.mocha.describe('Set lead proposal scenario', async () => {
+  initConfig();
+  registerJoystreamTypes();
+
+  const m1KeyPairs: KeyringPair[] = new Array();
+  const m2KeyPairs: KeyringPair[] = new Array();
+  const leadKeyPair: KeyringPair[] = new Array();
+
+  const keyring = new Keyring({ type: 'sr25519' });
+  const N: number = +process.env.MEMBERSHIP_CREATION_N!;
+  const paidTerms: number = +process.env.MEMBERSHIP_PAID_TERMS!;
+  const nodeUrl: string = process.env.NODE_URL!;
+  const sudoUri: string = process.env.SUDO_ACCOUNT_URI!;
+  const K: number = +process.env.COUNCIL_ELECTION_K!;
+  const greaterStake: BN = new BN(+process.env.COUNCIL_STAKE_GREATER_AMOUNT!);
+  const lesserStake: BN = new BN(+process.env.COUNCIL_STAKE_LESSER_AMOUNT!);
+  const applicationStake: BN = new BN(process.env.WORKING_GROUP_APPLICATION_STAKE!);
+  const roleStake: BN = new BN(process.env.WORKING_GROUP_ROLE_STAKE!);
+  const firstRewardInterval: BN = new BN(process.env.LONG_REWARD_INTERVAL!);
+  const rewardInterval: BN = new BN(process.env.LONG_REWARD_INTERVAL!);
+  const payoutAmount: BN = new BN(process.env.PAYOUT_AMOUNT!);
+  const alteredPayoutAmount: BN = new BN(process.env.ALTERED_PAYOUT_AMOUNT!);
+  const stakeDecrement: BN = new BN(process.env.STAKE_DECREMENT!);
+  const slashAmount: BN = new BN(process.env.SLASH_AMOUNT!);
+  const durationInBlocks: number = 70;
+
+  const provider = new WsProvider(nodeUrl);
+  const apiWrapper: ApiWrapper = await ApiWrapper.create(provider);
+  const sudo: KeyringPair = keyring.addFromUri(sudoUri);
+
+  setTestTimeout(apiWrapper, durationInBlocks);
+  membershipTest(apiWrapper, m1KeyPairs, keyring, N, paidTerms, sudoUri);
+  membershipTest(apiWrapper, m2KeyPairs, keyring, N, paidTerms, sudoUri);
+  membershipTest(apiWrapper, leadKeyPair, keyring, 1, paidTerms, sudoUri);
+  councilTest(apiWrapper, m1KeyPairs, m2KeyPairs, keyring, K, sudoUri, greaterStake, lesserStake);
+
+  let createOpeningProposalId: BN;
+  let openingId: BN;
+  tap.test(
```

Comment:
I see now the mechanism you mentioned at one point for getting results out of a test. It works well and I suppose its the only way really. But it can be improved. I think we need to do a better approach of encapsulating the calls, context and results of executing them to make things cleaner.

What would be really nice to see is something like this:

```javascript

const context = { apiWrapper, keys, ... }
const params = { applicationStake, roleStake, ... } 
const expectations = { expectFailure, expectedValue/Events }

// asserts that state of runtime is what is expected for execution and prepares the execution to run
// in a tap.test call
const createOpeningFixture = new CreateWorkingGroupLeaderOpeningFixture(context, params)

// runs and asserts on expectations
tap.test('opening', createOpeningFixture.runner(expectations))

// at this point the createOpeningFixture has within it the result of execution, events, the porposalId for example.
// that can be used as inputs for the following tests.
createOpeningFixture.result 
createOpeningFixture.events
createOpeningFixture.openingId 
```

This is what I would have really liked to see as a pattern throughout. This makes constructing and writing scenarios very reusable.

Reply:
Made a note for a future refactoring in Iznik